build.sc

```diff
@@ -17,7 +17,7 @@ import com.github.lolgab.mill.mima._
 val scala212  = "2.12.18"
 val scala213  = "2.13.11"
 
-val scala3   = "3.3.3"
+val scala3   = "3.4.2"
```

Comment:
any reason for this?

Reply:
Yeah, 3.3.3 fails with this exception

```scala
[error]     |Exception occurred while executing macro expansion.
[error]     |java.lang.ClassCastException: class dotty.tools.dotc.ast.Trees$This cannot be cast to class dotty.tools.dotc.ast.Trees$RefTree (dotty.tools.dotc.ast.Trees$This and dotty.tools.dotc.ast.Trees$RefTree are in unnamed module of loader mill.api.ClassLoader$$anon$1 @4f12ea86)
[error]     |	at scala.quoted.runtime.impl.QuotesImpl.scala$quoted$runtime$impl$QuotesImpl$reflect$Ref$$$_$apply$$anonfun$6(QuotesImpl.scala:446)
[error]     |	at scala.quoted.runtime.impl.QuotesImpl$reflect$.scala$quoted$runtime$impl$QuotesImpl$reflect$$$withDefaultPos(QuotesImpl.scala:3002)
[error]     |	at scala.quoted.runtime.impl.QuotesImpl$reflect$Ref$.apply(QuotesImpl.scala:446)
[error]     |	at scala.quoted.runtime.impl.QuotesImpl$reflect$Ref$.apply(QuotesImpl.scala:444)
[error]     |	at upickle.implicits.macros.macros$package$.apply$1(macros.scala:221)
[error]     |	at upickle.implicits.macros.macros$package$.applyConstructorImpl(macros.scala:255)
```

This is the code in question

```scala
    val companion: Symbol = tpe.classSymbol.get.companionModule
    val constructorParamSymss = tpe.typeSymbol.primaryConstructor.paramSymss

    def isPrimaryApplyMethod(syms1: Seq[Seq[Symbol]], syms2: Seq[Seq[Symbol]]) = {
      // try to guess the primary apply method based on the parameter counts
      // not sure why comparing the types doesn't seem to work
      // println(syms1.flatten.zip(syms2.flatten).map{case (s1, s2) => (s1.typeRef.simplified, s2.typeRef.simplified)})
      syms1.map(_.length) == syms2.map(_.length)
    }

    val applyMethods = companion
      .methodMember("apply")
      .filter(s => isPrimaryApplyMethod(s.paramSymss, constructorParamSymss))

    val lhs = Select(Ref(companion), applyMethods.head)
```

It's the `Ref.apply` call that is failing. It was beyond my abilities to debug why, but somehow going onto 3.4.2 fixes it

Reply:
ok, seems to be because of https://github.com/scala/scala3/issues/19732, fixed in 3.4

Reply:
it might not be easy to workaround that in 3.3.3 (like reimplementing Ref.apply)

Reply:
oh it looks like the fix is backported to 3.3.4-RC1